eio/client/dns_client_eio.ml

```diff
@@ -0,0 +1,204 @@
+type 'a env = <
+  clock         : Eio.Time.clock ;
+  net           : Eio.Net.t ;
+  fs            : Eio.Fs.dir Eio.Path.t ;
+  secure_random : Eio.Flow.source;
+  ..
+> as 'a
+
+type io_addr = Ipaddr.t * int
+type stack = {
+  sw            : Eio.Switch.t;
+  clock         : Eio.Time.clock;
+  net           : Eio.Net.t;
+  resolv_conf   : Eio.Fs.dir Eio.Path.t
+}
+
+module Transport : Dns_client.S 
+  with type io_addr = io_addr
+   and type stack   = stack
+   and type +'a io  = 'a  
+= struct 
+  type nonrec io_addr = io_addr
+  type nonrec stack = stack
+  type +'a io = 'a
+  type context = <Eio.Net.stream_socket; Eio.Flow.close>
+
+  type nameservers = 
+    | Static of io_addr Queue.t
+    | Resolv_conf of {
+        mutable nameservers : io_addr Queue.t ;
+        mutable digest      : Digest.t option ;
+      }
+  
+  type t = {
+    nameservers : nameservers ;
+    timeout_ns  : int64 ;      (* Timeout in nano seconds *)
+    stack       : stack;
+    mutex       : Eio.Mutex.t ;
+  }
+
+  let read_file file =
+    match Eio.Path.load file with
+    | content -> Ok content
+    | exception e ->
+      Fmt.error_msg "Error while reading file %a: %a" Eio.Path.pp file Fmt.exn e
+
+  (* Prioritises IPv6 nameservers before IPv4 nameservers so that we 
+     are more conformant with the happy eyballs RFC when implementing it. 
+     https://datatracker.ietf.org/doc/html/rfc8305#section-3  *)
```

Comment:
I'm curious, why not just use the happy-eyeballs library, as done in dns-client for lwt and mirage?

Reply:
For now all I wanted to do was sort name-server addresses in IP6 order. However, this package `dns-client-eio` will be required to implement `happy-eyeballs-eio`. So I want to leave happy-eyeballs implementation to the latter package.

Reply:
The happy-eyeballs package is independent of dns -> the happy-eyeballs-lwt/mirage/eio are dependent on dns-client. The pure happy-eyeballs is especially crafted for the use in dns, it allows to deal with ipv4 and ipv6 addresses. Please take a look into dns-client-lwt/dns-client-mirage, there's no reason to ad-hoc "sort name-server addresses in IP6 order".

Reply:
Ack. I will look into using happy-eyeballs for this part.